napari/layers/labels/labels.py

```diff
@@ -1142,9 +1148,21 @@ def get_color(self, label):
         elif label is None or (
             self.show_selected_label and label != self.selected_label
         ):
-            col = self.colormap.map([0, 0, 0, 0])[0]
+            col = self.colormap.map([self._background_label])[0]
         else:
-            col = self.colormap.map([label])[0]
+            raw_dtype = self._slice.image.raw.dtype
+            view_dtype = self._slice.image.view.dtype
+            if (
+                label < 0
+                and self.color_mode == LabelColorMode.AUTO
+                and raw_dtype.itemsize <= 2
+            ):
+                # This is casting for the data of type int8 and int16
+                # to have the same color as in canvas
+                val = np.array([label]).astype(view_dtype)[0]
```

Comment:
I guess we don't need the if clause... the conversion should be a no-op in other cases? 🤔

Reply:
Nope. 
This is just reproduced logic from `_cast_labels_to_minimum_dtype_auto` where `itemsize <= 2` is converted to proper uint and for bigger values the modulo operation is performed. 

also `view_dtype` is either uint8  or uint16 that will not filt all data from uint32, int64 etc.

Reply:
Hmmmm. 🤔 Ok. But I don't like the phrase "reproduced logic". 😅 Can we think of a way to just reuse that code path? I think the principle needs to be: every Labels value needs to go through the same code path from data to colour, with the only allowed differences being whether the code is executed on the CPU or GPU. I realise that this might not be achievable in just this PR, but it would be good to not add divergent branches...